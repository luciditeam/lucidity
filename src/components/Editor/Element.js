import React from 'react'
import {connect} from 'cerebral/react'

export default connect(
  ({elementRef}) => ({element: `editor.composition.paragraphs.${elementRef}`}),
  function Element ({element}) {
    const type = element.type

    const onInput = e => {
      console.log(e.type, e.target)
    }

    const onKeyPress = e => {
      console.log(e.key)
      switch (e.key) {
        case 'Enter':
          return e.preventDefault()
        default:
          // do nothing
      }
    }

    switch (type) {
      case 'p':
        return <p data-ref={element.ref} onKeyPress={onKeyPress} onInput={onInput} contentEditable dangerouslySetInnerHTML={{__html: element.text}} />
      default:
        return null
    }
  }
)
