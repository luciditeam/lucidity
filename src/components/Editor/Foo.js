import React from 'react'
import {connect} from 'cerebral/react'
import {Computed} from 'cerebral'

const Name = Computed(
  {
    name: 'foo.bar'
  },
  function Name ({name}) {
    return name
  }
)

export default connect(
  {
    name: Name
  },
  function Foo ({name}) {
    return <div>{name}</div>
  }
)
