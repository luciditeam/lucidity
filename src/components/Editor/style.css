.Editor {
  padding: 20px;
  background: #ddd;
  border: 1px solid #aaa;
}

.Editor *:focus {
  outline: none;
}
