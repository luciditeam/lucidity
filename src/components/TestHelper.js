import React from 'react'
import {StateContainer} from 'cerebral/react'
import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'

export function snapshot (block, state) {
  const wrapper = mount(
   <StateContainer state={state}>
      <div id='testhelper'>
       {block}
     </div>
   </StateContainer>
 )

  return mountToJson(wrapper.find('#testhelper').children())
}
