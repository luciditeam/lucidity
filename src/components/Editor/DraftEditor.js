import React from 'react'
import {Editor as Draft, EditorState} from 'draft-js'
import {stateToMarkdown} from 'draft-js-export-markdown'
import {connect} from 'cerebral/react'

import './style.css'

let state = EditorState.createEmpty()

export default connect(
  {
    editorState: 'editor.$draft.**'
  },
  {
    onChange: 'editor.draftChanged'
  },
  function Editor ({onChange}) {
    const change = c => {
      const content = c.getCurrentContent()
      console.log(stateToMarkdown(content))
      onChange({key: 'foo', value: Math.random()})
      state = c
    }
    return <div className='Editor'>
      <Draft editorState={state} onChange={change} />
    </div>
  }
)
