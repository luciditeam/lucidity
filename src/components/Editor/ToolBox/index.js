import React from 'react'
import {connect} from 'cerebral/react'

import './style.css'

export default connect(
  null,
  function ToolBox () {
    return (
      <div className='ToolBox-wrapper'
        contentEditable={false}
        >
        <div className='ToolBox'>
          <div className='ToolBox-menu'>
            <div className='ToolBox-item'>
              <i className='heading'>H1</i>
            </div>
            <div className='ToolBox-item'>
              <i className='heading'>H2</i>
            </div>
            <div className='ToolBox-item'>
              <i className='heading'>H3</i>
            </div>
            <div className='ToolBox-item'>
              <i className='para'>P</i>
            </div>
          </div>
        </div>
      </div>
    )
  }
)
