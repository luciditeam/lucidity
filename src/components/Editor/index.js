/* global Node */
import React from 'react'
import {connect} from 'cerebral/react'
import expandInner from './lib/expandInner'
import getPath from './lib/getPath'
import getSelection from './lib/getSelection'
import './style.css'

import ToolBox from './ToolBox'

export default connect(
  {
    compositionInner: 'editor.composition.i.*'
  },
  {
    inputChange: 'editor.inputChanged',
    backspacePress: 'editor.backspacePressed',
    enterPress: 'editor.enterPressed'
  },
  function Editor ({compositionInner, backspacePress, enterPress, inputChange}) {
    const onInput = e => {
      const {anchorNode} = window.getSelection()
      const path = getPath(anchorNode)
      const value = anchorNode.textContent
      inputChange({path, value})

      if (anchorNode.nodeType !== Node.TEXT_NODE) {
        // Not an edit
        return
      }
    }

    const onKeyDown = e => {
      switch (e.key) {
        case 'Enter':
          e.preventDefault()
          enterPress({selection: getSelection()})
          return
        case 'Backspace':
          e.preventDefault()
          backspacePress({selection: getSelection()})
          return
        default:
          // do nothing
      }

      const selection = getSelection()
      if (selection.hasSelection) {
        backspacePress({selection})
        // then continue with edit ?
      }
    }

    const onSelect = e => {
      const selection = getSelection()
      if (selection.anchorOffset === 0) {
        if (selection.anchorValue === '') {
          console.log('PARA.NEW', selection)
        } else {
          console.log('PARA.START', selection)
        }
      }
    }

    return <div className='Editor'
        onInput={onInput}
        onSelect={onSelect}
        onKeyDown={onKeyDown}
        contentEditable
        suppressContentEditableWarning
        >
        {expandInner('editor.composition', compositionInner)}
        <ToolBox />
      </div>
  }
)
