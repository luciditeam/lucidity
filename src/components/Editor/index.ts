import React from 'react'
import {connect} from 'cerebral/react'

import Composition from './Composition'
import ToolBox from './ToolBox'

export default connect(
  null,
  function Editor ({editorRef}) {
    // This editorRef should be unique by <Editor /> tag.
    let ref = `splendid33${editorRef || ''}`
    return <div className='Editor'>
        <div className='Editor-wrapper' id={ref}>
          <Composition />
          <ToolBox editorId={ref}/>
        </div>
      </div>
  }
)
