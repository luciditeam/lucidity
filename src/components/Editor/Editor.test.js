/* global it expect describe */
import React from 'react'
import {snapshot, render} from '../TestHelper'
import Editor from './'

const editorState = {
  composition: {
    paragraphs: {
      one: {
        ref: 'one',
        position: 0.0,
        type: 'P',
        text: 'This is the first message. Hello blah bomgolo frabilou elma tec.',
        markup: {
          iahyx: {
            start: 18,
            end: 25,
            type: 'STRONG'
          }
        }
      },
      two: {
        ref: 'two',
        position: 2.0,
        type: 'P',
        text: 'This is the second <b>message</b>. Hello blah bomgolo frabilou elma tec.'
      },
      three: {
        ref: 'three',
        position: 1.0,
        type: 'P',
        text: 'This is the third **message**. Hello blah bomgolo frabilou elma tec.'
      }
    }
  }
}

describe('editor', () => {
  it('renders without crashing', () => {
    render(<Editor />, {editor: editorState})
  })

  it('renders paragraphs in order', () => {
    const tree = render(<Editor />, {editor: editorState})
    expect(tree.find('Element').map(e => e.prop('elementRef'))).toEqual(['one', 'three', 'two'])
  })

  it('sets ref', () => {
    const tree = render(<Editor />, {editor: editorState})
    .find('[data-ref="one"]')
    expect(snapshot(tree)).toMatchSnapshot()
  })
})
