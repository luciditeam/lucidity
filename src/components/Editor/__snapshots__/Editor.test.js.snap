exports[`editor sets ref 1`] = `
<CerebralWrapping_Editor>
  <Editor
    backspacePress={[Function]}
    compositionInner={
      Object {
        "mcneu": Object {
          "i": Object {
            "jnaid": Object {
              "href": "http://example.com",
              "i": Object {
                "mnzjq": Object {
                  "i": "this ",
                  "p": 0,
                  "t": "T",
                },
                "zzvgp": Object {
                  "i": "link ",
                  "p": 1,
                  "t": "S+E",
                },
              },
              "p": 1,
              "t": "A",
            },
            "mnahl": Object {
              "i": "page",
              "p": 3,
              "t": "E",
            },
            "mznao": Object {
              "i": "to view the next ",
              "p": 2,
              "t": "T",
            },
            "ncgow": Object {
              "i": ".",
              "p": 4,
              "t": "T",
            },
            "uasuf": Object {
              "i": "You can click ",
              "p": 0,
              "t": "T",
            },
          },
          "p": 0,
          "t": "P",
        },
        "zaahg": Object {
          "i": "This is the third paragraph. My tailor types fast.",
          "p": 2,
          "t": "P",
        },
        "zhaog": Object {
          "i": Object {
            "haiou": Object {
              "i": ". Hello blah bomgolo frabilou elma tec.",
              "p": 2,
              "t": "T",
            },
            "oaiue": Object {
              "i": "message",
              "p": 1,
              "t": "S",
            },
            "oiafg": Object {
              "i": "This is the first ",
              "p": 0,
              "t": "T",
            },
          },
          "p": 1,
          "t": "P",
        },
      }
    }
    contentChange={[Function]}
    enterPress={[Function]}>
    <div
      className="Editor"
      contentEditable={true}
      onInput={[Function]}
      onKeyDown={[Function]}
      onSelect={[Function]}
      suppressContentEditableWarning={true}>
      <CerebralWrapping_Element
        elemRef="mcneu"
        path="editor.composition.i.mcneu">
        <Element
          elem={
            Object {
              "i": Object {
                "jnaid": Object {
                  "href": "http://example.com",
                  "i": Object {
                    "mnzjq": Object {
                      "i": "this ",
                      "p": 0,
                      "t": "T",
                    },
                    "zzvgp": Object {
                      "i": "link ",
                      "p": 1,
                      "t": "S+E",
                    },
                  },
                  "p": 1,
                  "t": "A",
                },
                "mnahl": Object {
                  "i": "page",
                  "p": 3,
                  "t": "E",
                },
                "mznao": Object {
                  "i": "to view the next ",
                  "p": 2,
                  "t": "T",
                },
                "ncgow": Object {
                  "i": ".",
                  "p": 4,
                  "t": "T",
                },
                "uasuf": Object {
                  "i": "You can click ",
                  "p": 0,
                  "t": "T",
                },
              },
              "p": 0,
              "t": "P",
            }
          }
          elemRef="mcneu"
          path="editor.composition.i.mcneu">
          <p
            className={null}
            data-ref="mcneu">
            <CerebralWrapping_Element
              elemRef="uasuf"
              path="editor.composition.i.mcneu.i.uasuf">
              <Element
                elem={
                  Object {
                    "i": "You can click ",
                    "p": 0,
                    "t": "T",
                  }
                }
                elemRef="uasuf"
                path="editor.composition.i.mcneu.i.uasuf">
                <span
                  className={null}
                  data-ref="uasuf">
                  You can click 
                </span>
              </Element>
            </CerebralWrapping_Element>
            <CerebralWrapping_Element
              elemRef="jnaid"
              path="editor.composition.i.mcneu.i.jnaid">
              <Element
                elem={
                  Object {
                    "href": "http://example.com",
                    "i": Object {
                      "mnzjq": Object {
                        "i": "this ",
                        "p": 0,
                        "t": "T",
                      },
                      "zzvgp": Object {
                        "i": "link ",
                        "p": 1,
                        "t": "S+E",
                      },
                    },
                    "p": 1,
                    "t": "A",
                  }
                }
                elemRef="jnaid"
                path="editor.composition.i.mcneu.i.jnaid">
                <a
                  className={null}
                  data-ref="jnaid">
                  <CerebralWrapping_Element
                    elemRef="mnzjq"
                    path="editor.composition.i.mcneu.i.jnaid.i.mnzjq">
                    <Element
                      elem={
                        Object {
                          "i": "this ",
                          "p": 0,
                          "t": "T",
                        }
                      }
                      elemRef="mnzjq"
                      path="editor.composition.i.mcneu.i.jnaid.i.mnzjq">
                      <span
                        className={null}
                        data-ref="mnzjq">
                        this 
                      </span>
                    </Element>
                  </CerebralWrapping_Element>
                  <CerebralWrapping_Element
                    elemRef="zzvgp"
                    path="editor.composition.i.mcneu.i.jnaid.i.zzvgp">
                    <Element
                      elem={
                        Object {
                          "i": "link ",
                          "p": 1,
                          "t": "S+E",
                        }
                      }
                      elemRef="zzvgp"
                      path="editor.composition.i.mcneu.i.jnaid.i.zzvgp">
                      <span
                        className="strong em"
                        data-ref="zzvgp">
                        link 
                      </span>
                    </Element>
                  </CerebralWrapping_Element>
                </a>
              </Element>
            </CerebralWrapping_Element>
            <CerebralWrapping_Element
              elemRef="mznao"
              path="editor.composition.i.mcneu.i.mznao">
              <Element
                elem={
                  Object {
                    "i": "to view the next ",
                    "p": 2,
                    "t": "T",
                  }
                }
                elemRef="mznao"
                path="editor.composition.i.mcneu.i.mznao">
                <span
                  className={null}
                  data-ref="mznao">
                  to view the next 
                </span>
              </Element>
            </CerebralWrapping_Element>
            <CerebralWrapping_Element
              elemRef="mnahl"
              path="editor.composition.i.mcneu.i.mnahl">
              <Element
                elem={
                  Object {
                    "i": "page",
                    "p": 3,
                    "t": "E",
                  }
                }
                elemRef="mnahl"
                path="editor.composition.i.mcneu.i.mnahl">
                <span
                  className="em"
                  data-ref="mnahl">
                  page
                </span>
              </Element>
            </CerebralWrapping_Element>
            <CerebralWrapping_Element
              elemRef="ncgow"
              path="editor.composition.i.mcneu.i.ncgow">
              <Element
                elem={
                  Object {
                    "i": ".",
                    "p": 4,
                    "t": "T",
                  }
                }
                elemRef="ncgow"
                path="editor.composition.i.mcneu.i.ncgow">
                <span
                  className={null}
                  data-ref="ncgow">
                  .
                </span>
              </Element>
            </CerebralWrapping_Element>
          </p>
        </Element>
      </CerebralWrapping_Element>
      <CerebralWrapping_Element
        elemRef="zhaog"
        path="editor.composition.i.zhaog">
        <Element
          elem={
            Object {
              "i": Object {
                "haiou": Object {
                  "i": ". Hello blah bomgolo frabilou elma tec.",
                  "p": 2,
                  "t": "T",
                },
                "oaiue": Object {
                  "i": "message",
                  "p": 1,
                  "t": "S",
                },
                "oiafg": Object {
                  "i": "This is the first ",
                  "p": 0,
                  "t": "T",
                },
              },
              "p": 1,
              "t": "P",
            }
          }
          elemRef="zhaog"
          path="editor.composition.i.zhaog">
          <p
            className={null}
            data-ref="zhaog">
            <CerebralWrapping_Element
              elemRef="oiafg"
              path="editor.composition.i.zhaog.i.oiafg">
              <Element
                elem={
                  Object {
                    "i": "This is the first ",
                    "p": 0,
                    "t": "T",
                  }
                }
                elemRef="oiafg"
                path="editor.composition.i.zhaog.i.oiafg">
                <span
                  className={null}
                  data-ref="oiafg">
                  This is the first 
                </span>
              </Element>
            </CerebralWrapping_Element>
            <CerebralWrapping_Element
              elemRef="oaiue"
              path="editor.composition.i.zhaog.i.oaiue">
              <Element
                elem={
                  Object {
                    "i": "message",
                    "p": 1,
                    "t": "S",
                  }
                }
                elemRef="oaiue"
                path="editor.composition.i.zhaog.i.oaiue">
                <span
                  className="strong"
                  data-ref="oaiue">
                  message
                </span>
              </Element>
            </CerebralWrapping_Element>
            <CerebralWrapping_Element
              elemRef="haiou"
              path="editor.composition.i.zhaog.i.haiou">
              <Element
                elem={
                  Object {
                    "i": ". Hello blah bomgolo frabilou elma tec.",
                    "p": 2,
                    "t": "T",
                  }
                }
                elemRef="haiou"
                path="editor.composition.i.zhaog.i.haiou">
                <span
                  className={null}
                  data-ref="haiou">
                  . Hello blah bomgolo frabilou elma tec.
                </span>
              </Element>
            </CerebralWrapping_Element>
          </p>
        </Element>
      </CerebralWrapping_Element>
      <CerebralWrapping_Element
        elemRef="zaahg"
        path="editor.composition.i.zaahg">
        <Element
          elem={
            Object {
              "i": "This is the third paragraph. My tailor types fast.",
              "p": 2,
              "t": "P",
            }
          }
          elemRef="zaahg"
          path="editor.composition.i.zaahg">
          <p
            className={null}
            data-ref="zaahg">
            This is the third paragraph. My tailor types fast.
          </p>
        </Element>
      </CerebralWrapping_Element>
      <CerebralWrapping_ToolBox>
        <ToolBox>
          <div
            className="ToolBox-wrapper"
            contentEditable={false}>
            <div
              className="ToolBox">
              <div
                className="ToolBox-menu">
                <div
                  className="ToolBox-item">
                  <i
                    className="heading">
                    H1
                  </i>
                </div>
                <div
                  className="ToolBox-item">
                  <i
                    className="heading">
                    H2
                  </i>
                </div>
                <div
                  className="ToolBox-item">
                  <i
                    className="heading">
                    H3
                  </i>
                </div>
                <div
                  className="ToolBox-item">
                  <i
                    className="para">
                    P
                  </i>
                </div>
              </div>
            </div>
          </div>
        </ToolBox>
      </CerebralWrapping_ToolBox>
    </div>
  </Editor>
</CerebralWrapping_Editor>
`;
